File: workers/python/python/tasker_core/registry/base_resolver.py
"""TAS-93: Abstract base class for step handler resolvers.

This module defines the contract that all resolvers must implement.
Resolvers are tried in priority order by the ResolverChain until
one successfully resolves the handler.

Resolution Contract:
1. name - Human-readable identifier for logging/debugging
2. priority - Lower numbers = tried first (10 = explicit, 100 = inferential)
3. can_resolve() - Quick check if this resolver might handle the definition
4. resolve() - Actually resolve the handler callable

Example Implementation:
    class MyResolver(BaseResolver):
        @property
        def name(self) -> str:
            return "my_custom_resolver"

        @property
        def priority(self) -> int:
            return 50  # Between explicit (10) and class lookup (100)

        def can_resolve(self, definition: HandlerDefinition, config: dict) -> bool:
            return definition.callable.startswith("MyApp.")

        def resolve(self, definition: HandlerDefinition, config: dict) -> Any | None:
            # Return handler or None
            pass
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from .handler_definition import HandlerDefinition


class BaseResolver(ABC):
    """Abstract base class for step handler resolvers.

    Defines the contract that all resolvers must implement.
    Resolvers are tried in priority order by the ResolverChain.
    """

    @property
    @abstractmethod
    def name(self) -> str:
        """Human-readable name for this resolver (for logging/debugging).

        Returns:
            The resolver name.
        """
        ...

    @property
    @abstractmethod
    def priority(self) -> int:
        """Resolution priority (lower = tried first).

        Standard priorities:
        - 10: Explicit mapping (registered handlers)
        - 100: Class lookup (inferential)

        Returns:
            The priority value.
        """
        ...

    @abstractmethod
    def can_resolve(
        self, definition: HandlerDefinition, config: dict[str, Any] | None = None
    ) -> bool:
        """Quick eligibility check (called before resolve).

        This should be a fast check to determine if this resolver
        is even worth trying for the given definition.

        Args:
            definition: Handler configuration.
            config: Additional configuration context.

        Returns:
            True if this resolver might be able to resolve the handler.
        """
        ...

    @abstractmethod
    def resolve(
        self, definition: HandlerDefinition, config: dict[str, Any] | None = None
    ) -> Any | None:
        """Resolve the handler callable from the definition.

        Args:
            definition: Handler configuration.
            config: Additional configuration context.

        Returns:
            Handler object or None if cannot resolve.
        """
        ...

    def registered_callables(self) -> list[str]:
        """Return all callable keys this resolver knows about.

        Used for debugging and introspection.

        Returns:
            List of registered callable identifiers.
        """
        return []

File: workers/ruby/lib/tasker_core/registry/resolvers/base_resolver.rb
# frozen_string_literal: true

module TaskerCore
  module Registry
    module Resolvers
      # TAS-93: Abstract base class for step handler resolvers
      #
      # This class defines the contract that all resolvers must implement.
      # Resolvers are tried in priority order by the ResolverChain until
      # one successfully resolves the handler.
      #
      # == Resolution Contract
      #
      # 1. `name` - Human-readable identifier for logging/debugging
      # 2. `priority` - Lower numbers = tried first (10 = explicit, 100 = inferential)
      # 3. `can_resolve?` - Quick check if this resolver might handle the definition
      # 4. `resolve` - Actually resolve the handler callable
      #
      # == Implementation Example
      #
      #   class MyResolver < BaseResolver
      #     def name
      #       'my_custom_resolver'
      #     end
      #
      #     def priority
      #       50  # Between explicit (10) and class constant (100)
      #     end
      #
      #     def can_resolve?(definition, config)
      #       # Quick eligibility check
      #       definition.callable.start_with?('MyApp::')
      #     end
      #
      #     def resolve(definition, config)
      #       # Return handler or nil
      #       # handler must respond to effective_method
      #     end
      #   end
      #
      class BaseResolver
        # Human-readable name for this resolver (for logging/debugging)
        #
        # @return [String] The resolver name
        def name
          raise NotImplementedError, "#{self.class}#name must be implemented"
        end

        # Resolution priority (lower = tried first)
        #
        # Standard priorities:
        # - 10: Explicit mapping (registered handlers)
        # - 100: Class constant (inferential)
        #
        # @return [Integer] The priority value
        def priority
          raise NotImplementedError, "#{self.class}#priority must be implemented"
        end

        # Quick eligibility check (called before resolve)
        #
        # This should be a fast check to determine if this resolver
        # is even worth trying for the given definition.
        #
        # @param definition [TaskerCore::Types::HandlerDefinition] Handler configuration
        # @param config [Hash] Additional configuration context
        # @return [Boolean] true if this resolver might be able to resolve the handler
        def can_resolve?(definition, config)
          raise NotImplementedError, "#{self.class}#can_resolve? must be implemented"
        end

        # Resolve the handler callable from the definition
        #
        # @param definition [TaskerCore::Types::HandlerDefinition] Handler configuration
        # @param config [Hash] Additional configuration context
        # @return [Object, nil] Handler object or nil if cannot resolve
        def resolve(definition, config)
          raise NotImplementedError, "#{self.class}#resolve must be implemented"
        end

        # Return all callable keys this resolver knows about (for debugging/introspection)
        #
        # @return [Array<String>] List of registered callable identifiers
        def registered_callables
          []
        end
      end
    end
  end
end
